### DISCUSSION TITLE
Has someone already created a CASL decorator?

### DISCUSSION THREAD

**Discussion by angelhodar** [MichalLytek/type-graphql#1444] [Ideas]
TypeGraphQL already has a built-in ``Authorized`` decorator that is great to use on RBAC. However, I miss some field based control to allow/disallow users to read/write certain row fields without proper permissions. For example, a user should not be able to update its ``balanceCredits`` field, but an admin could do it. I am thinking about creating a decorator that uses the great [casl library](https://casl.js.org/v6/en/) that could be used as the ``Authorized`` one, but I have doubts about what would be the performance cost of implementing it on a per field basis because the main use would be to filter fields on response based on user role

**Comment by carlocorradini** [comment]
Hey, I've created [`graphql-auth-directive`](https://github.com/carlocorradini/graphql-auth-directive) which can be used with TypeGraphQL.
Let me know if it works for your context :)

**Reply by angelhodar** [reply]
Hi Carlo, first of all thanks for your response, I think you have the record for the fastest answer I have received ever hahaha. I have seen your library and it is almost exactly what I was looking for because I just need to change your Auth decorator parameters with just a function that returns a casl ability instance with all the permissions. I just have doubts in this example you have in the typegraphql folder:

```ts
@Mutation(() => Post, { nullable: true })
  @Auth({
    roles: [UserRoles.ADMIN],
    permissions: [UserPermissions.DELETE_POST]
  })
  deletePost(@Arg('id', () => GraphQLInt) id: number) {
    return removePost(id);
  }
  ```
  
 What if I want to allow the user to delete his posts ONLY if the id parameter is from a post that the user owns? If you want I can open an issue in your repo instead of talking here. Greator work btw!

**Reply by carlocorradini** [reply]
Sorry for the late response.
You must execute a DB query inside the `auth` function/class.
The `auth` function/class has two parameters:
1. [`resolverData`](https://github.com/carlocorradini/graphql-auth-directive/blob/main/src/types/ResolverData.ts)
1. [`authData`](https://github.com/carlocorradini/graphql-auth-directive/blob/main/src/types/AuthData.ts)

You can use `resolverData` to obtain the `id` (from `resolverData.args.id`), then perform a DB query to obtain the owner of the object and check it against `authData` or the `user` in the current `context`.

**Reply by carlocorradini** [reply]
Is that what you were trying to achieve?

**Reply by angelhodar** [reply]
Hi Carlo! Yes that was what I was trying to figure out before the resolver execution. So my only doubt now is that in cases where the decorator is used with a prop, the ``resolverData`` should contain the info about the prop? For example:

```ts
class Example{
   @Auth({
    roles: [UserRoles.ADMIN, UserRoles.USER]
    permissions: [UserPermissions.READ]
  })
   @Field()
   restrictedField: string
}
```

What if an ADMIN user can always see the field but another user with the USER role can only see the field ``restrictedField`` depending on its value? And what is the performance costs of prop decorators? How are they executed?

**Reply by carlocorradini** [reply]
Since this is custom logic you must implement it in your custom `auth` function.
The value of `restrictedField` can be read in pure `graphql.js` (see [this](https://www.apollographql.com/docs/apollo-server/data/resolvers)):
```ts
import type { AuthFn } from 'graphql-auth-directive'

export const myAuthFn: AuthFn<MyContext, MyRole, MyPermission> = ({ source }, { roles, permissions }) => {
  const value = source.restrictedField;

  // ...
};
```

Of course, if you want to have the directive to be compatible with both `field` and "`functions`" you must check for the correcteness (and have a more complicated logic) but this works :)

**Reply by angelhodar** [reply]
I would create Restricted decorator for props and Auth for resolvers so I can keep it more simple. And in general what is the performance cost of this directives?

**Reply by carlocorradini** [reply]
@angelhodar I have no idea :/
Feel free to send a PR 😎🥳
